### ISSUE TITLE
Provide module extensions with a way to reference extension repos

### ISSUE THREAD

**Issue by fmeum** [bazelbuild/bazel#19055]
### What underlying problem are you trying to solve with this feature?

The repository mapping that applies to a module extension does not include the repositories it generates. This makes it very tricky to pass a reference to some file in such a repo to another repo rule instantiated by the same extension, e.g.:

```starlark
def _ext_impl(module_ctx):
    my_lang_toolchain_repo(
        name = "lang_sdk",
    )
    my_toolchain_repo(
        name = "lang_toolchain",
        # This does *not* work as `lang_sdk` is not in the current file's repo mapping.
        sdk_root = "@lang_sdk//:root",
    )
```

This is a very common pattern, but requires major hacks to pull of without additional API: 
1. Observe that the canonical name of an extension repo is `<common prefix>~<apparent name>` and rely on that implementation detail to construct the canonical name of `lang_sdk` in `lang_toolchain` as `ctx.attr.name.removesuffix("lang_toolchain") + "lang_sdk"`. 
2. Make use of the fact that `lang_sdk` is in the repo mapping of the repositories generated by the extension. By making `sdk_root` an `attr.string` instead of an `attr.label` and generating a `.bzl` file in some other ("hub") repo in the the same extension that calls `Label` on a label string, `lang_toolchain` can pass that literal through the correct repo mapping. This requires the module to `use_repo` the hub repo and is overall pretty complex.

### Description of the feature request:

Provide a function on `module_ctx` that converts the name of a repository generated by the extension and a repo-relative label into an instance of `Label` referencing the intended target.

This could look as follows:

```starlark
# In rules_foo v1.2.3, extensions.bzl
def _ext_impl(module_ctx):
    repo(name = "foo", ...)
    repo(name = "bar", ...)

    # Returns `Label("@rules_foo~1.2.3~ext~foo//dir:name")`
    module_ctx.extension_repo_label("@foo//dir:name")
    # Returns `Label("@rules_foo~1.2.3~ext~bar//:name")`
    module_ctx.extension_repo_label("@bar//:name")

    # Fails immediately, argument has to be absolute
    module_ctx.extension_repo_label("//:name")
    # Fails immediately, label must not be canonical
    module_ctx.extension_repo_label("@@rules_foo~1.2.3~ext~foo//dir:name")
    # Fails immediately, repo name must be a valid "user-specified" name
    module_ctx.extension_repo_label("@rules_foo~1.2.3~ext~foo//dir:name")
    # Fails after the module extension returns as `not_foo` is not generated by the extension.
    module_ctx.extension_repo_label("@not_foo//dir:name")
```

Alternative ideas for names: 
* `label_in_repo`

Alternative signatures: 
* Separate arguments for apparent repository name and repo-relative label.

Alternative approaches:
* Hook into the repo rule creation and defer the conversion of label strings into `Label` instances until after the module extension has fully evaluated, at which point the names of all the repos it generates are known. Then use this information to convert label strings implicitly. This has the advantage that the "obvious" approach just works and the downside that its implementation is probably more involved.

### Which operating system are you running Bazel on?

N/A

### What is the output of `bazel info release`?

6.3.0

**Comment by fmeum** [comment]
CC @Wyverald @meteorcloudy, I am especially interested in your opinion on the alternative approach. If it is feasible, I could see this being better than introducing new API.

**Comment by Wyverald** [comment]
The alternative approach might not be possible since you could do this:

```bzl
def _extension_impl(mctx):
  repo_rule(name="foo")
  print(str(Label("@foo//:abc")))
  # or with the API
  print(mctx.extension_repo_label("@foo//:abc"))
```

**Comment by fmeum** [comment]
Do we really want users to construct `Label`s into extension repos though? If they feed them to `module_ctx.path`, that would probably always result in cycles or other kinds of failures. Supporting label attributes of repo rules is what this is really about and I have a hard time coming up with any other usage that would make sense. But I agree that if we can think of just one then we should just expose a function that is guaranteed to handle all cases.

**Comment by meteorcloudy** [comment]
It's interesting to see how rules_java works around this:
https://github.com/bazelbuild/rules_java/blob/0878192f194292421788d35bea53533b0c730aec/toolchains/remote_java_repository.bzl#L86

It basically passes the label (`"@{repo}//:jdk".format(repo = name)`) as a plain string to generate the BUILD file, so that the label actually got resolved during build time. This works if the label isn't needed in the repo rule implementation function of `my_toolchain_repo`, which I don't know is often true or false.

**Comment by Wyverald** [comment]
I think it's a bit too late for such an API addition at this point. We can revisit for 7.1.0.

**Comment by fmeum** [comment]
Snippets in rulesets that rely on cutting and pasting canonical repo names:
1. https://github.com/bazelbuild/bazel-gazelle/blob/852fdcfe7d6cdfcc72b1275a2200a1e131bf0203/internal/go_repository.bzl#L312-L320 (this would be much better if it could go through the existing `ctx.attr.build_config` `attr.label`)
2. https://github.com/bazelbuild/rules_python/pull/1625/files/cb916380cc36654030aa93ef28281cd4052a2bda#diff-4d74a8d7b92c579455fb1a4142f10748eb4d217d1c0b767db151422785aace3bR132-R136 (not merged, this one would store the label in a variable rather than pass it to a repo rule directly)

**Comment by fmeum** [comment]
I'm leaning towards resolving labels in repo rule arguments against an augmented repo mapping that contains all other repos created by the extension. If you aren't an expert on repo mappings, this is actually what I would take "A repo generated by a module extension can see all repos visible to the module that hosts the extension, plus all other repos generated by the same module extension (using the names specified in the repo rule calls as their apparent names)." (taken from the current docs) to mean. 

`Label` manipulation within the module extension wouldn't pick up these mappings, but I think that's fine: If it would, it would pretty much always result in a cycle.

**Comment by Wyverald** [comment]
> I'm leaning towards resolving labels in repo rule arguments and implementation functions against an augmented repo mapping that contains all other repos created by the extension.

I'm not sure if that's feasible. At the time a repo rule is invoked to create a repo, we already need to give it the repo mapping it needs to convert string labels to label objects (https://cs.opensource.google/bazel/bazel/+/master:src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionEvalStarlarkThreadContext.java;drc=300c5867b7d2da1ba32abc20e95662096c2a7a08;l=110). This at least means that we can't write `repo(name='foo', lbl='@bar')` before `repo(name='bar')`.

So the flip side is that we _can_ make `repo(name='bar'); repo(name='foo', lbl='@bar')` work. The catch is that we need to constantly update the repo mapping object as we create repos in an extension impl function. This can result in a _lot_ of temporary garbage for an extension that creates thousands of repos (`RepositoryMapping` uses an `ImmutableMap` right now). Thankfully it's just temporary garbage as we don't store the `Rule` objects representing the repos anymore, just the `RepoSpec`s.

> If you aren't an expert on repo mappings, this is actually what I would take "A repo generated by a module extension can see all repos visible to the module that hosts the extension, plus all other repos generated by the same module extension (using the names specified in the repo rule calls as their apparent names)." (taken from the current docs) to mean.

I know you're an expert on repo mappings (😁), but I still felt the need to call out that "the mapping of the extension creating the repos" is not the same as "the mapping of the repos being created". And the mapping of the extension creating the repos is what's used to, well, create the repos. I do realize it's subtle and thus can be unintuitive to non-experts...

**Comment by Wyverald** [comment]
> This at least means that we can't write `repo(name='foo', lbl='@bar')` before `repo(name='bar')`.

Thinking about it again, we could actually achieve this by deferring all repo creation to the end of the extension impl. Whenever a repo rule is invoked, we just note down the name and the attributes without parsing them or converting them. Then at the end we actually create the repos en masse.

This obviously means that errors will be reported late, but it's not later by much so whatever.

Another potential pitfall is the case where the module hosting the extension has a `bazel_dep` on `foo`, while the extension also generates a repo called `foo`. Right now, the string label `@foo` will refer to the `bazel_dep` (as the generated repo `foo` is not part of the repo mapping). If we do the magic thing, then the semantics will switch around.

I'm kind of coming around to the idea, but I wonder if we should still have an explicit API anyways, just for the case where someone _does_ want a label referring into a just-created repo for any reason.

**Comment by fmeum** [comment]
> Thinking about it again, we could actually achieve this by deferring all repo creation to the end of the extension impl. Whenever a repo rule is invoked, we just note down the name and the attributes without parsing them or converting them. Then at the end we actually create the repos en masse.

Yes, this is what I had in mind. Since we don't have `native.existing_rule(s)`, repo creation doesn't have observable non-fatal side effects. This is even closer to the more functional API you described to me a while ago where we would return a list of repo "values" to create repos instead of calling repo "functions" that create them as a side effect.

> This obviously means that errors will be reported late, but it's not later by much so whatever.
They will be reported late in time (the repo rule may take a while), but as we can also keep track of the locations at which the repo was instantiated, we should be able to generate the same error message. There is a small caveat: If the module extension implementation encounters some other error after instantiating a repo with invalid attributes, that error would now be reported first. We *could* mostly prevent that by collecting all repos, running until completion or the first evaluation error and, in case we hit an error, evaluating all repo rules created up to that point and throwing *their* errors first before reporting the evaluation error. Not sure whether that's worth it though.
> Another potential pitfall is the case where the module hosting the extension has a bazel_dep on foo, while the extension also generates a repo called foo. Right now, the string label @foo will refer to the bazel_dep (as the generated repo foo is not part of the repo mapping). If we do the magic thing, then the semantics will switch around.

It would be a change in behavior, but no behavior seems clearly better than the other. Also, there would still be a way to refer to the `bazel_dep` in a label attribute, namely by wrapping it in `Label`. I guess that we could also decide to keep the old behavior, but that would be even more magical (as the "repo mapping" we would end up emulating would differ from both the one of the extension and the one of the repos we create).If we change the behavior, how about just making this an error? I can only see such a collision to be fine if the extension doesn't use the `bazel_dep` at all. If a downstream run doesn't result in any failures, I would consider this the safest choice for Bazel 8.

> I'm kind of coming around to the idea, but I wonder if we should still have an explicit API anyways, just for the case where someone does want a label referring into a just-created repo for any reason.

Do you see any (even if contrived) use case for this in which there isn't an alternative that's clearly better? A relevant reason for why I prefer the magic API is that I currently consider making this impossible a *feature*.

**Comment by Wyverald** [comment]
> Not sure whether that's worth it though.

I don't think it's worth doing :)


> Also, there would still be a way to refer to the `bazel_dep` in a label attribute, namely by wrapping it in `Label`.

This would be a very subtle way to work around a very subtle problem, but at least it is a way to disambiguate. I don't like the situation, but I also don't have a better idea, so this is fine by me.



> Do you see any (even if contrived) use case for this in which there isn't an alternative that's clearly better? A relevant reason for why I prefer the magic API is that I currently consider making this impossible a _feature_.

No, nothing comes to mind.

**Comment by Wyverald** [comment]
@bazel-io fork 7.4.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!
